[GSoC'24] M1.4: Show classroom name in topic editor and avoid deletion and unpublish operations

AFZL210 opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #19849 .
2. This PR does the following: Add topic name in the topic editor page and sorting by classroom name. Also, only delete a topic if it is not assigned to a classroom. If the topic is assigned to a classroom, show a warning asking topic managers to contact curriculum admins and remove the topic from the classroom first.

- Show classroom name and curriculum admin usernames in topic editor page if topic is assigned to a classroom [Figma mock](https://www.figma.com/design/VVcmDcYGvGeJGAJs6TzuQP/Classroom-Mocks?node-id=150-8&t=9gopbLoOXCi5JyAb-4)

- Only delete a topic if it's not assigned to a classroom. If it is assigned then show a warning saying "Contact curriculum admins and remove it from X classroom first"

- Only unpublish a topic if it's not assigned to a classroom. If it is assigned then show a warning saying "Contact curriculum admins and remove it from X classroom first"
- Topics can be sorted by classroom name and changed default sorting order to "Most recently updated"

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [ ] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct

Sort by classroom name

https://github.com/oppia/oppia/assets/79896602/469bca21-254c-43ef-b829-e713ebda5523



Show classroom name in topic editor and avoid unpublish operation if topic is assigned to a classroom


https://github.com/oppia/oppia/assets/79896602/ba8e0f3c-29c4-421e-935f-ef3b22ea6224



Show warning message when deleting topic which is assigned to a classroom


https://github.com/oppia/oppia/assets/79896602/b21d84c9-0f4e-4578-a0fd-eae6632ee6bb




<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network
N/A
<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone
N/A
<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language
N/A
<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

AFZL210 committed:
Fix part of #19849: Display classroom name in topic editor

AFZL210 committed:
fix deleteTopic

AFZL210 committed:
fix classroom type

Nik-09 reviewed (changes requested):
Hi @AFZL210 Thank you for the PR. Can you please first get approval from the primary mentor since this PR is part of the GSoC project?
Let me know once it is done.

Thank you

AFZL210 committed:
update sorting order to most recently updated

AFZL210 committed:
Merge branch 'develop' into fix#19849-classroom-topic-editor-page

seanlip commented:
@AFZL210 I'm not exactly sure why you assigned me, since I'm not codeowner of any files ... is it because you'd like a product review? If so please state that explicitly next time.

In general I think the functionality looks good, just a couple questions/issues:

- In the first video, shouldn't the default sort order be by last-updated?
- In the second video, please have the "contact curriculum admin" help text be similar to the help text for the fields above it (e.g. "Meta tags provide information ..."). Currently it's too prominent.

Thanks!

AFZL210 commented:
> @AFZL210 I'm not exactly sure why you assigned me, since I'm not codeowner of any files ... is it because you'd like a product review? If so please state that explicitly next time.
> 
> In general I think the functionality looks good, just a couple questions/issues:
> 
> * In the first video, shouldn't the default sort order be by last-updated?
> * In the second video, please have the "contact curriculum admin" help text be similar to the help text for the fields above it (e.g. "Meta tags provide information ..."). Currently it's too prominent.
> 
> Thanks!

Yes just for the product review.

1) I will update the video
2) Sure will update this too

Thank you!

AFZL210 committed:
Merge branch 'develop' into fix#19849-classroom-topic-editor-page

AFZL210 committed:
updated curriculum admin contact text

AFZL210 committed:
Merge branch 'develop' into fix#19849-classroom-topic-editor-page

AFZL210 committed:
add test for getClassroomName()

AFZL210 committed:
fix backend test

AFZL210 commented:
@nikitaevg I have updated the video proof too. PTAL. Thank you!

nikitaevg commented on core/controllers/topic_editor.py:
```diff
@@ -541,14 +544,29 @@ def get(self, topic_id: str) -> None:
         skill_creation_is_allowed = (
             role_services.ACTION_CREATE_NEW_SKILL in self.user.actions)
 
+        curriculum_admin_usernames = (
```
admin -> admins, not a serious issue, but "admin_usernames" sounds like one admin might have multiple usernames

AFZL210 replied:
yes make sense but in other parts like in email_manger.py we just use admin. So should I change it or it's fine?

nikitaevg replied:
Consistency is a good argument, so let's keep it then, nice observation!

nikitaevg commented on core/controllers/topic_editor.py:
```diff
@@ -541,14 +544,29 @@ def get(self, topic_id: str) -> None:
-            'classroom_url_fragment': classroom_url_fragment,
+            'classroom_url_fragment': (
+                None if (
+                    str(constants.CLASSROOM_URL_FRAGMENT_FOR_UNATTACHED_TOPICS)
```
Why `str`? It's already a string, isn't it?

AFZL210 replied:
yes it is but in the runtime the type is not know (it's any) as its extracted from a ts file.

nikitaevg commented on core/controllers/topic_editor.py:
```diff
@@ -541,14 +544,29 @@ def get(self, topic_id: str) -> None:
+            'classroom_url_fragment': (
+                None if (
+                    str(constants.CLASSROOM_URL_FRAGMENT_FOR_UNATTACHED_TOPICS)
+                    == classroom_url_fragment
```
Almost always it's \<variable\> == \<const\>, not \<const\> == \<variable\>.

AFZL210 replied:
Done.

nikitaevg commented on core/domain/classroom_config_services.py:
```diff
@@ -165,6 +165,22 @@ def get_classroom_url_fragment_for_topic_id(topic_id: str) -> str:
+    for classroom in classrooms:
+        if topic_id in classroom.get_topic_ids():
+            return classroom.name
+    return str(constants.CLASSROOM_NAME_FOR_UNATTACHED_TOPICS)
```
I don't get this. Why not return None?

AFZL210 replied:
The learner dashboard part is based on this fact that topic which are not assigned to any classroom are in "staging" mode. So it might break that part.

nikitaevg replied:
Ok, thanks for explaining

nikitaevg commented on core/templates/pages/topic-editor-page/editor-tab/topic-editor-tab.directive.ts:
```diff
@@ -236,8 +239,11 @@ export class TopicEditorTabComponent implements OnInit, OnDestroy {
-  getClassroomUrlFragment(): string {
-    return this.topicEditorStateService.getClassroomUrlFragment();
+  isAssignedToAClassroom(): boolean {
+    if (this.classroomUrlFragment && this.classroomName) {
```
"return this.classroomUrlFragment && this.classroomName" is good enough

AFZL210 replied:
Done.

AFZL210 replied:
Done.

nikitaevg commented on core/templates/pages/topics-and-skills-dashboard-page/topics-list/topics-list.component.ts:
```diff
@@ -115,11 +115,36 @@ export class TopicsListComponent {
   deleteTopic(topicId: string, topicName: string): void {
+    const topicSummary = this.getTopicSummaryForTopicId(topicId);
+
+    if (topicSummary && topicSummary.classroom) {
```
if (topicSummary?.classroom) should work?

AFZL210 replied:
Done.

AFZL210 committed:
refactor code

AFZL210 committed:
fix acceptance test for deleting a topic

AFZL210 commented:
@nikitaevg PTAL. Thanks!

AFZL210 committed:
fix acceptance test for deleting a topic

nikitaevg commented on core/controllers/topic_editor.py:
```diff
@@ -541,14 +544,31 @@ def get(self, topic_id: str) -> None:
         self.values.update({
-            'classroom_url_fragment': classroom_url_fragment,
+            'classroom_url_fragment': (
+                None if (
```
Don't you here put None effectively, just like I suggested below? Won't it break the learner dashboard?

AFZL210 replied:
No it won't as the endpoint is used only in the creator facing part which is topic and skill dashboard. and we don't really need to make any changes to anything related to learner dashboard part.

AFZL210 committed:
fix test

AFZL210 committed:
fix test

AFZL210 commented:
@DubeySandeep PTAL. I have updated e2e test for topicAndSkillsDashboard in the delete topic part as we cannot delete topic which is assigned to a classroom so i have removed the step in which we assign the topic-one to math classroom just for delet test case.

AFZL210 committed:
Merge branch 'develop' into fix#19849-classroom-topic-editor-page

AFZL210 commented:
@DubeySandeep I have updated the e2e test for topic and skills dashboard as now we cannot delete topic which are assigned to a classroom. So in the test case of deleting a topic I have removed the classroom assignment step.

Nik-09 commented on core/controllers/topic_editor_test.py:
```diff
@@ -66,13 +66,21 @@ def setUp(self) -> None:
         self.save_new_skill(
             self.skill_id_2, self.admin_id, description='Skill Description 2')
         self.topic_id = topic_fetchers.get_new_topic_id()
+        self.topic_id_no_classroom = topic_fetchers.get_new_topic_id()
```
Can we rename this variable? Currently, this name is very unclear.

AFZL210 replied:
yes sure. I guess topic_id_with_no_classroom makes sense

Nik-09 reviewed (changes requested):
Thank you @AFZL210, only 1 comment, PTAL.

AFZL210 committed:
update variable name

AFZL210 committed:
Merge branch 'develop' into fix#19849-classroom-topic-editor-page

Nik-09 reviewed (approved):
Thank @AFZL210, LGTM

AFZL210 committed:
Merge branch 'develop' into fix#19849-classroom-topic-editor-page
